Issue 1195 - Enable writing json as logmessage

DaveSchmid opened the pull request:
<!--- In the TITLE (↑↑↑↑ above ↑↑↑↑ **NOT HERE**) provide a general, short summary of your changes -->
<!--- Please DO NOT use references to other PR's or issues -->

## Description
<!--- Describe your changes in detail -->
<!--- Bulleted list. Full sentences. Ending with a dot. -->
String.Format in the logger message without any parameter arguments prevents writing Json-string as a log message.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If this **fixes** OR **closes** OR  **resolves** an open issue, please link to the issue there using the template bellow (mind the pattern to link there as all issues are tracked in the Home repository) -->
<!--- **JUST** replace NNNNN with the issue number -->
- Fixes/Closes/Resolves nanoFramework/Home#1195

## How Has This Been Tested?<!-- (IF APPLICABLE) -->
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested with an example. Unit tests don't work on my VS2022.
## Screenshots<!-- (IF APPROPRIATE): -->

## Types of changes
<!--- What types of changes does this PR introduce? Put an `x` in all the boxes that apply: -->
- [x ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- PLEASE PLEASE PLEASE don't tick all of them just because -->
- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.Logging?pullRequest=184) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.Logging?pullRequest=184) before we can accept your contribution.<br/><br/>:x: DaveSchmid [sign now](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.Logging?pullRequest=184)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nanoframework/nanoFramework.Logging?pullRequest=184) it.</sub>

josesimoes commented:
It's OK to add the Unit tests. If you still have them around pleae add.

josesimoes commented:
@DaveSchmid I've just updated the unit tests.  Please pull the changes. Close VS. Delete `.vs` folder, rebuild and hit run all. It should work now. VS2022 preferably.

DaveSchmid commented:
@josesimoes I got the latest code but that's not my problem. Whenever I try to start something from nanoframework (also with all the samples) it always has a problem with the references. I always have to run PM > Update-Package first in Packet.Manager otherwise it won't find the references and won't be able to compile. And it still doesn't work for the unit tests. I can see that there are tests but I can't debug or start them.

![2022-12-20_19-54-44](https://user-images.githubusercontent.com/146932/208750003-7dca92c5-1006-4f7f-b4e0-2ffa7fb55ebd.png)

![2022-12-20_20-03-20](https://user-images.githubusercontent.com/146932/208750037-ed84e876-84d9-41b4-a457-e541d4b22272.png)

DaveSchmid commented:
ok, it works in VS2019 !  Not in VS2022 and not loadable in VS2017.

josesimoes commented:
Is that on VS 2022?

nano project system isn't able to perform nuget restore, therefore you have to do it manually from the solution explorer.

Sometimes VS can get a bit nuts about this. The remedy is :
1. close VS
2. manually delete bin, obj and .vs folders
3. open again, restore
4. rebuild

That should get you up and running. Sometimes simply closing and opening the solution after the restart is enough.

DaveSchmid commented:
maybe it's also a problem with my NuGet package sources. I still have nanoframework packages from
https://pkgs.dev.azure.com/nanoframework/feed/_packaging/sandbox/nuget/v3/index.json
Should I better delete this?

josesimoes commented:
We've moved to stable releases for quite some time. So yes, you can get rid of the preview feed.

DaveSchmid commented:
OK. I found out something about the broken references! Actually it has nothing to do with this PR but I'm writing it here anyway as an answer to my problem. When I freshly get a project from nanoFramework I always have to run Update-Package in Packet-Manager first or set the property "Specific Version" in the reference of  mscorlib to "No". Only then is compiling possible. A manual NuGet restore is not helping! It says everything is fine, but that's not true. I think this is an error for an expert?

Ellerbach commented on nanoFramework.Logging/LoggerExtensions.cs:
```diff
@@ -410,7 +410,11 @@ public static void Log(this ILogger logger, LogLevel logLevel, EventId eventId,
-            logger.Log(logLevel, eventId, string.Format(message, args), exception, MessageFormatter);
+            string formatedMessage = message; 
+            if(args.Length > 0) 
+                formatedMessage = string.Format(message, args);
```
We're using explicit brackets in nanoFramework. 

 ```suggestion
       {
                formatedMessage = string.Format(message, args);
        }
```

Ellerbach replied:
Other wise, you can even do like:
```csharp
string formatedMessage = args.Length > 0 ? string.Format(message, args) : message;
```

DaveSchmid replied:
The shortest solution with explicit brackets would be this:

![2022-12-30_13-04-09](https://user-images.githubusercontent.com/146932/210088283-6096f106-2403-424a-8888-6bfe8148d97f.png)
It would change the reference from the method declaration without using a new variable. On my old PC and ESP32 over Com-Port and with enabled "step filtering messages"  in the debugger, it takes hours for the debugger to jump to the next block of code. There are small facts that are far more important than visual style, which is technically irrelevant.
I personally find it much more important how the code behaves in the debugger. And there is a big difference between if/else and ?: . Even if you break a block with ternary operator, the debugger can't. It may be possible sometime in the future, but it's not possible at the moment.

![2022-12-23_09-53-18](https://user-images.githubusercontent.com/146932/210088310-1dd98430-11d7-42e1-b13b-5568e0d21069.png)

![2022-12-23_09-58-53](https://user-images.githubusercontent.com/146932/210088337-1bf05d5a-1f57-49ed-b219-73d480bfacd0.png)

![2022-12-23_09-54-39](https://user-images.githubusercontent.com/146932/210088348-4e1c5553-164d-4e16-b77c-ef1384c2b638.png)

![2022-12-23_09-56-13](https://user-images.githubusercontent.com/146932/210088363-3d099299-37b1-4271-a06d-7ec00f9b1282.png)

But the most important thing is unfortunately missing from this review.
It's something that can be discussed endlessly, but my feeling about it is very clear:
The code must be documented! And the right place is in the code. No one would know what this code is for unless it is described. Of course it's also good if there is a test that checks this and a requirement call for it and description somewhere in the source control and in a Spec., but you should see it when you look at the code!

PS: The reason why I didn't check in a test is that the test framework doesn't work properly for me. Not even with the newest Christmas versions. It's acting super strange! There is still a lot of work to be done before it works the way it should.

Ellerbach reviewed (commented):
Comment about formating as we are using explicit bracket. And a shortest solution proposed!

josesimoes commented:
Not really sure what is bugging at your end, but I can run unit tests all day here without any issues...

We love PRs feel absolutely free to send those with improvements. 😁

josesimoes commented:
Regarding the fact that "it takes hours" to execute a couple of lines in debug that's probably because you're _stepping into_ library code. It takes one (or more) round trips to the execution engine to execute those.
This is executing in the mcu, it goes back and forth in a serial connection. Don't expect it to have a performance even close to what you have when debugging a desktop application.
The solution it's much simple: don't step into code that's not "yours" 😉

josesimoes commented:
By all means: add as much verbose comments as you like.

DaveSchmid commented:
> Not really sure what is bugging at your end, but I can run unit tests all day here without any issues...
> 
> We love PRs feel absolutely free to send those with improvements. 😁

Please set some brackepoints on line 43, 50, 58 of LoggerExtensionsTests.cs

josesimoes commented:
It's not possible to debug Uni Tests libraries in nanoFramework. Only run them.

DaveSchmid commented:
why? I can't program without it. I can't even do it with

Ellerbach commented:
> why? I can't program without it. I can't even do it with

You can't debug the unit tests but you can run them. The reason why you can't debug them is because the running is an application itself that is deployed and the tests are deployed as well with this specific application. And the way VS gather the results does not allow any debug. To write units tests to test all this, it should be very straight forward. You can get inspiration with those in the Json library for example: https://github.com/nanoframework/nanoFramework.Json/tree/main/nanoFramework.Json.Test

Ellerbach commented:
When you select one of the tests, you have all the output as well available, everything that is output, exceptions, errors, everything.

DaveSchmid commented:
ok tests are green! 
(Not so happy with this testing framework.)

josesimoes commented:
> ok tests are green! 

Yes. 


> (Not so happy with this testing framework.)

It's the one we have 🤷🏻‍♂️ ... On the "outside" it's the same as MSTest V2. On the inside the adapter and the runner had to be updated for nanoFramework, of course. It took a while a significant amount of work to pull it of, it wasn't even possible before VS2019. 
Like I've mentioned before: I 💜 PRs so, if you have any improvements to offer for the test framework, by all means, let's have those.

DaveSchmid commented:
@josesimoes I have great respect for your work. I tried to analyze the TestFramework a bit because certain things don't work as expected. But it's too complicated for me. There's a lot I have no idea about. And actually I want to program something different and just use this framework as a base.

Ellerbach commented:
@DaveSchmid let me take your PR, add the tests so you'll see ;-) I'll push things back here (I'm on slow internet, so it can take a bit of time)

DaveSchmid commented:
> @DaveSchmid let me take your PR, add the tests so you'll see ;-) I'll push things back here (I'm on slow internet, so it can take a bit of time)

the tests are already there. What they don't check is the output from the logger. Unfortunately there is no Debug.ReadLine()

Ellerbach reviewed (approved):
I've made few adjustments, simplified the setup for the logger, added few tests and adjusted the code a bit while adding the tests. Thanks again for the help. I really don't know what went wrong for the tests for you, just worked out of the box for me!

Ellerbach commented:
@josesimoes I let you merge it, there is a rate limit hit on the version check. I guess those are the proper versions. You can double check anyway.

DaveSchmid commented:
> I've made few adjustments, simplified the setup for the logger, added few tests and adjusted the code a bit while adding the tests. Thanks again for the help. I really don't know what went wrong for the tests for you, just worked out of the box for me!

Exactly that with the "static ILogger _logger;" didn't work for me! Why? I have weird problems.

josesimoes commented:
I want to clean up the solution and the nfproj first. Several build configs were added that are not usable, like x64. These are breaking the pattern that we use everywhere so they need to be reverted.

DaveSchmid commented:
> x64

I added x64 to it because of error messages when running the tests. But I'm not sure if that's necessary.

Ellerbach commented:
> > I've made few adjustments, simplified the setup for the logger, added few tests and adjusted the code a bit while adding the tests. Thanks again for the help. I really don't know what went wrong for the tests for you, just worked out of the box for me!
> 
> Exactly that with the "static ILogger _logger;" didn't work for me! Why? I have weird problems.

You have to initialize it in a Setup function. That's the trick. The class is never created or loaded. Only the decorated functions are called. Nothing else. Checkout the Unit Test launcher: https://github.com/nanoframework/nanoFramework.TestFramework/blob/main/source/UnitTestLauncher/Program.cs

For simplicity and to avoid filling up the memory, you only call the functions. The `Setup` one(s) is(are) called first. The `Cleanup` at the end. The `TestMethod` in the middle. This pattern allows as well to make sure you can properly run some tests or skip them. In our logging, if you don't have a serial port or System.IO available to log files, the series of tests are skipped magically. This is happening in the `Setup` function.

Ellerbach commented:
> I want to clean up the solution and the nfproj first. Several build configs were added that are not usable, like x64. These are breaking the pattern that we use everywhere so they need to be reverted.

I let you do that, didn't checked this! I should have...

DaveSchmid commented:
> You have to initialize it in a Setup function. That's the trick. The class is never created or loaded. Only the decorated functions are called. Nothing else. Checkout the Unit Test launcher: https://github.com/nanoframework/nanoFramework.TestFramework/blob/main/source/UnitTestLauncher/Program.cs

It seems that the setup function is not always executed. I've already tried that but it didn't work. That's why I did it this way.

DaveSchmid commented:
The documentation for nanoFramework.TestPlatform should be supplemented that debugging is not possible. It made me very confused as a beginner to nanoFramework.
-> https://docs.nanoframework.net/content/unit-test/constraints-limitations.html
-> https://github.com/nanoframework/nanoFramework.TestFramework/blob/main/README.md

Ellerbach commented:
> The documentation for nanoFramework.TestPlatform should be supplemented that debugging is not possible. It made me very confused as a beginner to nanoFramework. -> https://docs.nanoframework.net/content/unit-test/constraints-limitations.html -> https://github.com/nanoframework/nanoFramework.TestFramework/blob/main/README.md

Please provide a PR, we ❤️ PR. This time, no test to add... Thanks again for your help, your patience and your feedbacks!
